beater/mux.go

```diff
@@ -0,0 +1,220 @@
+package beater
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/elastic/apm-server/processor"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"compress/gzip"
+	"compress/zlib"
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"crypto/subtle"
+
+	"github.com/elastic/beats/libbeat/monitoring"
+)
+
+type chainable func(processor.Processor, Config, reporter) http.Handler
+
+var (
+	serverMetrics  = monitoring.Default.NewRegistry("apm-server.server")
+	requestCounter = monitoring.NewInt(serverMetrics, "requests.counter")
+	responseValid  = monitoring.NewInt(serverMetrics, "response.valid")
+	responseErrors = monitoring.NewInt(serverMetrics, "response.errors")
+
+	errInvalidToken    = errors.New("invalid token")
+	errPOSTRequestOnly = errors.New("only POST requests are supported")
+
+	handlerMap = map[int]chainable{
+		processor.Backend:  backendHandler,
+		processor.Frontend: frontendHandler,
+		processor.Nop:      nopHandler,
+	}
+)
+
+func newMuxer(config Config, report reporter) *http.ServeMux {
+	mux := http.NewServeMux()
+
+	for path, p := range processor.Registry.Processors() {
+		handler := handlerMap[p.Type()]
+		logp.Info("Path %s added to request handler", path)
+		mux.Handle(path, handler(p, config, report))
+	}
+
+	return mux
+}
+
+func backendHandler(p processor.Processor, config Config, report reporter) http.Handler {
+	return logHandler(
+		authHandler(config.SecretToken,
+			appHandler(p, config, report)))
+}
+
+func frontendHandler(p processor.Processor, config Config, report reporter) http.Handler {
+	return logHandler(
+		appHandler(p, config, report))
+}
+
+func nopHandler(_ processor.Processor, _ Config, _ reporter) http.Handler {
+	return logHandler(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			sendStatus(w, r, 200, nil)
+		}))
+}
+
+func logHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logp.Debug("handler", "Request: URI=%s, method=%s, content-length=%d", r.RequestURI, r.Method, r.ContentLength)
+		requestCounter.Inc()
+		h.ServeHTTP(w, r)
+	})
+}
+
+func authHandler(secretToken string, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !isAuthorized(r, secretToken) {
+			sendStatus(w, r, 401, errInvalidToken)
+			return
+		}
+		h.ServeHTTP(w, r)
+	})
+}
+
+// isAuthorized checks the Authorization header. It must be in the form of:
+//   Authorization: Bearer <secret-token>
+// Bearer must be part of it.
+func isAuthorized(req *http.Request, secretToken string) bool {
+	// No token configured
+	if secretToken == "" {
+		return true
+	}
+	header := req.Header.Get("Authorization")
+	parts := strings.Split(header, " ")
+	if len(parts) != 2 || parts[0] != "Bearer" {
+		return false
+	}
+	return subtle.ConstantTimeCompare([]byte(parts[1]), []byte(secretToken)) == 1
+}
+
+func appHandler(p processor.Processor, config Config, report reporter) http.Handler {
```

Comment:
maybe `processHandler` is a more appropriate name here. It does the actual processing.

Reply:
or `processRequestHandler`